### PR TITLE
feature(js): Adds hooks to pass site and page-level data client-side

### LINKED ISSUE #8997: Document configuring AMD modules via [config, amd]
This lets us sneak server-side data to modules via the elgg/require_config module. The idea is to eliminate the need for PHP modules just to pass a bit of server data in.

### PR DESCRIPTION
**Pull request by mrclay**
Adds two hooks `"elgg.data"` that populate the object `elgg.data` available in the `elgg` module.

Fixes #8997

### CODE REVIEW

**Comment by mrclay** [comment]
@hypeJunction thoughts on this?

**Comment by hypeJunction** [comment]
In Iceland till the end of the week. Will review when I am back from
holiday.

File: docs/guides/hooks-list.rst [hunk 1]
```diff
@@ -115,6 +115,12 @@ System hooks
 
 **add, river**
 
+**elgg.data, site**
```
Comment (hypeJunction) [inline docs/guides/hooks-list.rst]: Not convinced by hook names. Maybe `'js:data','site'`?

Comment (hypeJunction) [inline docs/guides/hooks-list.rst]: I suppose you want them to be consistent with JS naming convention, but that also bring discrepancy to hook names we have server-side

File: views/default/elgg.js.php [hunk 2]
```diff
@@ -51,11 +51,13 @@
...
- */
+// Set some values that are cacheable
+$data = (object)elgg_trigger_plugin_hook('elgg.data', 'site', null, []);
 ?>
 //<script>
+// page data overrides site data
+elgg.data = $.extend(<?= json_encode($data) ?>, elgg._data);
+delete elgg._data;
 
 elgg.version = '<?php echo elgg_get_version(); ?>';
```
Comment (hypeJunction) [inline views/default/elgg.js.php]: Can we move these into PHP array that is passed to the hook, or merged with the hook return value? Would be nice to clean up the spaghetti code below.

Comment (hypeJunction) [inline views/default/elgg.js.php]: Maybe add `_elgg_get_init_client_site_data()` and rename the other to `_elgg_get_init_client_page_data()`?

**Comment by hypeJunction** [comment]
LGTM. The comments are negligible if you don't have time for time. One thing I am curious about is when would you recommend using these hooks versus AMD config hook?

**Comment by mrclay** [comment]
> when would you recommend using these hooks versus AMD config hook?

Good question. The AMD hook can also push config data into particular modules, though we haven't documented any of that. We should probably do that first and see if we're happy with that. Only case the AMD hook doesn't cover is simple uses of `require()` outside modules.

**Comment by hypeJunction** [comment]
Let's rebase and get this in. There is a need for a it and using AMD config is undocumented and a bit convoluted.

File: docs/guides/hooks-list.rst (hunk 1 above)
Comment (mrclay) [inline docs/guides/hooks-list.rst]: How about: `js:elgg.data, site`?

Comment (mrclay) [inline docs/guides/hooks-list.rst]: Or `register, elgg.data:site`

Comment (hypeJunction) [inline docs/guides/hooks-list.rst]: Now that I look at it, it doesn't really matter

File: engine/lib/views.php [hunk 3]
```diff
@@ -1801,6 +1801,80 @@ function _elgg_manage_pagesetup($hook, $view, $value, $params) {
...
+ *
+ * @return array
+ * @access private
+ */
+function _elgg_get_js_page_data() {
+	$data = elgg_trigger_plugin_hook('elgg.data', 'page', null, []);
+	if (!is_array($data)) {
+		elgg_log('"elgg.data" plugin hook handlers must return an array. Returned ' . gettype($data) . '.', 'ERROR');
+		$data = [];
+	}
```
Comment (mrclay) [inline engine/lib/views.php]: We now complain if hook returns not an array.

**Pull request merged**